### PR TITLE
FIX: 1-d-numpy and pd input for gradient explainer

### LINKED ISSUE #3152: BUG: gradient Explainer pd.DataFrame input failing
### Issue Description

The gradient explainer class accepts `pandas.Dataframe` as a data input for TF models. The dataframe is correctly used for its feature_name columns, but is then passed on as a `pd.Dataframe` to `_TFGradient`, which does not deal with this type properly.
A simple fix would be to convert the Dataframe to its underlying np.array

### Minimal Reproducible Example

```python
import shap
    import pandas as pd
    import tensorflow as tf

    feature_names = [f"feature_{i}" for i in range(5)]

    background = pd.DataFrame(np.zeros((10, 5)), columns=feature_names)

    model = tf.keras.Sequential([
        tf.keras.layers.Dense(10, input_shape=(5,), activation='relu'),
        tf.keras.layers.Dense(1, activation='linear')
    ])
    model.compile(optimizer='adam', loss='mse')
    e = shap.GradientExplainer(model, background)
```

### Expected Behavior

The `GradientExplainer` should deal with pd.Dataframes as input as is described.

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [X] I'd be interested in making a PR to fix this bug

### Installed Versions

0.42.1

### PR DESCRIPTION
**Pull request by ponseko**
## Overview

Closes #3152, #2607 (#3025)

Description of the changes proposed in this pull request:

#3152 is fixed by simply converting a pd.Dataframe to its underlying np.array in `_TFGradient` (after setting the feature names in `Gradient`).

#2607 (#3025) seemed stale, so I figured I'd close that by simply setting the feature_names to None if there is no pd.Dataframe provided. It seems like the labels are set by the plots if the feature names are None, so there is no need to set them to a list of numbers here. 
Additionaly, an extra parameter can be added here for the feature names (like some explainers already do). But I am unsure whether, going forward, you would want to keep that logic in the plots themselves only.

Added two unit tests for simple pd and np 1-d inputs.

Let me know if anything needs changing.

## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: tests/explainers/test_gradient.py [hunk 1]
```diff
@@ -271,3 +271,62 @@ def forward(self, x1, x2):
     sums = np.array([shap_x1[i].sum() + shap_x2[i].sum() for i in range(len(shap_x1))])
     d = np.abs(sums - diff).sum()
     assert d / (np.abs(diff).sum()+0.01) < 0.1, "Sum of SHAP values does not match difference! %f" % (d / np.abs(diff).sum())
+
+def test_tf_pd_input(random_seed):
+    """ Test tabular (batch_size, features) pd.DataFrame input. """
+    pd = pytest.importorskip('pandas')
```
Comment (connortann) [inline tests/explainers/test_gradient.py]: Pandas is a non-optional dependency of the package, so it can be safely imported in the normal way at the top of the file

File: tests/explainers/test_gradient.py [hunk 2]
```diff
@@ -271,3 +271,62 @@ def forward(self, x1, x2):
...
+    # testing pd input
+    e = shap.GradientExplainer(model, background)
+    shap_values = e.shap_values(x1[:1])
+
+    diff = (model.predict(x1[:1]) - model.predict(background.values)).mean(0)
+    sums = np.array([shap_values[i].sum() for i in range(len(shap_values))])
+    d = np.abs(sums - diff).sum()
+    assert d / (np.abs(diff).sum() + 0.01) < 0.1, "Sum of SHAP values does not match difference! %f" % (d / np.abs(diff).sum())
+
+def test_tf_np_input(random_seed):
```
Comment (connortann) [inline tests/explainers/test_gradient.py]: The two unit tests have a fair amount of duplicated code, so I would suggest combining them and using pytest's `parameterize` to provide different different model inputs

**Review by connortann — CHANGES_REQUESTED** [review]
Thanks for the PR, this looks like a great improvement!

I've added a couple of minor suggestions for changes to the unit tests.

File: tests/explainers/test_gradient.py [hunk 3]
```diff
@@ -271,3 +271,62 @@ def forward(self, x1, x2):
     sums = np.array([shap_x1[i].sum() + shap_x2[i].sum() for i in range(len(shap_x1))])
     d = np.abs(sums - diff).sum()
     assert d / (np.abs(diff).sum()+0.01) < 0.1, "Sum of SHAP values does not match difference! %f" % (d / np.abs(diff).sum())
+
+def test_tf_pd_input(random_seed):
+    """ Test tabular (batch_size, features) pd.DataFrame input. """
+    pd = pytest.importorskip('pandas')
+    tf = pytest.importorskip('tensorflow')
+    import pandas as pd
+    import tensorflow as tf
```
Comment (thatlittleboy) [inline tests/explainers/test_gradient.py]: `pytest.importorskip` returns the imported module itself, so there's no need to import again below.

**Comment by ponseko** [comment]
@thatlittleboy @connortann Thank you for the feedback. I've included the requested changes to the unit tests in the latest commit.

File: tests/explainers/test_gradient.py [hunk 4]
```diff
@@ -271,3 +272,38 @@ def forward(self, x1, x2):
...
+
+    e = shap.GradientExplainer(model, background)
+    shap_values = e.shap_values(x1[:1])
+
+    if input_type == "dataframe":
+        diff = (model.predict(x1[:1]) - model.predict(background.values)).mean(0)
+    elif input_type == "numpy":
+        diff = (model.predict(x1[:1]) - model.predict(background)).mean(0)
+    else:
+        raise NotImplementedError("Input type %s not implemented" % input_type)
```
Comment (connortann) [inline tests/explainers/test_gradient.py]: I think this `if/else` clause is unnecessary here, as this should work for both cases:

```python
diff = (model.predict(x1[:1]) - model.predict(background)).mean(0)
```

File: tests/explainers/test_gradient.py [hunk 5]
```diff
@@ -271,3 +272,38 @@ def forward(self, x1, x2):
...
+        background = pd.DataFrame(background, columns=feature_names)
+    x1 = np.ones((batch_size, num_features))
+
+    model = tf.keras.Sequential([
+        tf.keras.layers.Dense(10, input_shape=(num_features,), activation='relu'),
+        tf.keras.layers.Dense(1, activation='linear')
+    ])
+    model.compile(optimizer='adam', loss='mse')
+
+    e = shap.GradientExplainer(model, background)
```
Comment (connortann) [inline tests/explainers/test_gradient.py]: This test is in line with the rest of the unit tests in style; but I think we can improve on it a bit. We could use the newer `Explainer` API here, and use some slightly more descriptive variable names. Over time we could update the rest of the tests to match.

How about something like this?

```python
explainer = shap.GradientExplainer(model, background)
example = np.ones((1, num_features))
explanation = explainer(example)
diff = (model.predict(example) - model.predict(background)).mean(0)
sums = np.array([values.sum() for values in explanation.values])
```

File: tests/explainers/test_gradient.py (hunk 4 above)
Comment (ponseko) [reply tests/explainers/test_gradient.py]: Oops, I convinced myself the TF model required np input, but the predict function deals with that just fine.

**Review by connortann — APPROVED** [review]
This looks great - thanks for making those tweaks.
